### DISCUSSION TITLE
init()시점에 viewDidLoad()를 호출하는 것에 대한 궁금증이 있습니다.

### DISCUSSION THREAD

**Discussion by snowedev** [nsoojin/MiniSuperApp-fastcampus#12] [아무 질문이나 환영합니다] [answered]
안녕하세요, 수진님께 메일을 통해 질문을 드렸었는데 Discussion이 있는것을 이제 알았네요. 궁금한 점이 하나 더 생겨 질문드립니다. 

현재 프로젝트 코드는 viewDidLoad()가 아닌 init()시점에 addSubView()등을 해주고 있는데요,
이렇게 viewDidLoad()가 아니라 init()시점에 뷰에 접근하는 코드를 작성하게 되면 ViewController가 인스턴스화 됨과 동시에 ViewDidLoad()가 호출이 될 것 같습니다.

이와 관련하여 RIBs 구조에서의 예를 들면 저희가 build 메서드 내에서 이 뷰컨을 인스턴스화 해놓은 뒤, 필요할 때 attach, present를 하게 될텐데요, 
init()시점에 ViewDidLoad()까지 호출이 이루어진다면 뷰가 present되는 시점이 아닌 인스턴스화 되는 시점에 view를 전부 로드시킬 것이라고 생각됩니다. 이러한 경우 init()시점에 viewDidLoad()가 호출되는 것이 성능상의 비효율성을 갖진 않을지, 수진님께서는 어떻게 생각하고 계신지 궁금합니다!

**Comment by nsoojin** [comment, marked as answer]
안녕하세요.
생각해본적이 없는 흥미로운 질문이네요.
제대로된 답변을 드리기 전에 저도 질문이 몇개 생겼는데요,

1. 어떤 부분에서 비효율성이 발생할 수 있을지, 어떤 종류의 비효율성을 의미하신건지 자세히 알 수 있을까요? 
2. 빌더의 build와 뷰컨 present를 시간차를 두고 사용해본적이 없어서 언제 그런 경우가 있나요? 그런 경우가 많으신가요?

**Reply by snowedev** [reply]
RIBs를 사용하는 일반적인 패턴에서는 말씀드린 문제가 생기지 않을 것 같습니다. 

하지만 만약 특수한 상황에서 Router를 만들어두고 이동시켜야 하는 일이 있다고 할 때, attach 시점이 아닌 인스턴스화되는 시점에 모든 뷰가 만들어진다면 attach전까지 불필요한 낭비를 초래할 것 같습니다. 

또한 RIBs 아키텍쳐가 아닌 상황에서 탭바를 사용한다고 했을 때에도 만약 인스턴스화 시점에 viewDidLoad 내의 코드들이 호출이 된다면 뷰의 생성뿐만 아니라 서비스 호출 코드도 불릴 수 있을 것 같습니다. 이런경우에 불필요한 자원의 소모가 발생하지 않을까 생각하였습니다

직접 경험한 사이드이펙트를 한가지 말씀드리자면, 저는 RIBs와 ReactorKit을 함께 사용하고 있는데요. View에서 ReactorKit의 존재를 모르게 하기 위해서 View의 PresentableListener를 통해 구현을 해주었습니다. 위와같은 이유로 View에서는
```siwft
bind(listener: listener)
```
와 같이 리스너를 통해 State의 변화에 대한 바인딩 로직을 구현하였습니다.

이러한 경우 init시점에 ViewDidLoad가 didBecomeActive보다 먼저 호출된다면, ViewDidLoad시점에서 listener에 nil이 들어가게 됩니다. 따라서 bind가 제대로 이루어지않는 결과를 발생시키게 되더라구요.

위와같은 이유로 init()시점에 viewDidLoad()가 호출됨으로써 야기할 수 있는 여러 사이드이펙트가 있다고 생각하였습니다! 혹시 이해가 되셨을까요?

**Reply by nsoojin** [reply]
말씀하신 비효율이라는게 (1) view가 그려지기 전에 미리 인스턴스들이 생성되는것과, (2) viewDidLoad에서 불리는 어떤 자원을 사용하는 코드를 말씀하신거 같습니다. 

(1)번의 경우, 뷰들이 레이아웃되고 그려지는게 아니고 인스턴스만 생성되는 것인데, 뷰 객체 십 수개가 조금 더 일찍 생성되는 것이 성능을 저해시킬 이슈라거나, 낭비라고는 생각하기 힘듭니다. 필요할때 바로 쓰기 위해 미리 loading/initialize 해놓는 디자인 패턴도 있고요.
(2)번의 경우, 뷰컨트롤러를 뷰로 바라보는 RIBs 하에서는 viewDidLoad 메서드 안에서 비즈니스로직이 호출되면 안됩니다. 저는 리액터킷을 쓰지 않기 때문에 바인딩을 어떻게 하시는지 알지 못하지만, 뷰와 인터랙터를 바인딩해야한다면 didBecomeActive에서 해주는게 올바른 접근일거 같습니다. viewDidLoad에서 비즈니스 로직을 시작하는 건 MVC의 특성이라, 혼용해서 쓰면 RIBs의 장점을 누리기 힘들거 같습니다. RIBs를 쓸때는 뷰컨에서는 최대한 아무런 결정을 직접하는 일이 없이 시키는 일만 하고, 그리라는대로만 그리는 수동적인 객체로 만들기를 추천합니다.
